kvserver: improve tenant_id CPU observability

tbg opened the pull request:
Motivated by https://github.com/cockroachlabs/support/issues/3297.

This commit

- adds a new `replicas.cpunanospersecond` metric, which aggregates the Replica
ReqCPUNanosPerSecond at the tenant level.
- Adds a tenant_id tag to CPU profiles.

This should simplify investigations related to tenant-induced overload: the new
metric should often help pinpoint the set of hot tenants. CPU profiles can help
dig into the specific code paths this tenant is exercising. This can then be
rounded out with the existing metrics for request counts by tenant (both
received by KV and sent by the tenant Pod) for a comprehensive picture.

Release note: the `replicas.cpunanospersecond` metric was added. Notably, when
child labels are enabled, it exposes evaluation-related Replica CPU usage by
tenant.
Epic: none

tbg committed:
kvserver: improve tenant_id CPU observability

Motivated by https://github.com/cockroachlabs/support/issues/3297.

This commit

- adds a new `replicas.cpunanospersecond` metric, which aggregates the Replica
ReqCPUNanosPerSecond at the tenant level.
- Adds a tenant_id tag to CPU profiles.

This should simplify investigations related to tenant-induced overload: the new
metric should often help pinpoint the set of hot tenants. CPU profiles can help
dig into the specific code paths this tenant is exercising. This can then be
rounded out with the existing metrics for request counts by tenant (both
received by KV and sent by the tenant Pod) for a comprehensive picture.

Release note: the `replicas.cpunanospersecond` metric was added. Notably, when
child labels are enabled, it exposes evaluation-related Replica CPU usage by
tenant.
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146526)
<!-- Reviewable:end -->

tbg committed:
kvserver: nest tenant metrics ref inside of tenant metrics

In a follow-up step, we will absorb the struct into its parent, thus removing a
confusing layer of indirection. The Replica will hold on to a `*tenantMetrics`
directly.

tbg committed:
kvserver: absorb tenant metrics ref

This should be simpler overall: there is a single struct that represents the metrics for a tenant on the store.

tbg commented:
Had to add some preliminary cleanup commits to address data races - the correct use of tenant metrics is notoriously tricky because the metrics must not be used once the replica is destroyed (which could drop the tenant's refcount to zero, which would destroy the metrics object as well). It's really hard to synchronize cheaply with replica destruction, since we haven't made it straightforward.

I simplified the metrics lifecycle management so that it's now more obvious (hopefully) how this works. The new metric is then updated by acquiring and then releasing a reference on the metrics object. Instead of the race (which thankfully was caught in CI) the behavior would now be that after the metrics reference is released, we make a new one, make one bogus update, and release that one again. But this is all well-defined behavior that "looks" exactly like the normal lifecycle of the metrics, and it's a good way to handle this annoying special case.

I'll request a review once CI is happy.

stevendanna reviewed (commented):
> Reviewed 6 of 6 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146526)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/146526#-OT1ZB6ZEOV5ocLDty7H:-OT1ZB6ZEOV5ocLDty7I:bz7kpyb) (no related file):*
Nice! Will be interesting to see our first use of this. Just one small question.
___
*[`pkg/kv/kvserver/metrics.go` line 3409 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146526#-OT1XYTHA8EqcaKAKV4t:-OT1XYTHA8EqcaKAKV4u:b-lwbjap) ([raw file](https://github.com/cockroachdb/cockroach/blob/c860badd2cd9cfbca541eab370f8695397988360/pkg/kv/kvserver/metrics.go#L3409)):*
> ```Go
> 		refCount int
> 		released atomic.Bool // allowed to read without holding mu
> 		stack    debugutil.SafeStack
> ```

I don't see who is setting this now.


<!-- Sent from Reviewable.io -->

tbg committed:
kvserver: add missing stack assignment

This was lost in a previous refactor.

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146526)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[`pkg/kv/kvserver/metrics.go` line 3409 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146526#-OT1XYTHA8EqcaKAKV4t:-OTTSlQM9BZX7RM0HO9T:b-6qb4hu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c860badd2cd9cfbca541eab370f8695397988360/pkg/kv/kvserver/metrics.go#L3409)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I don't see who is setting this now.
</blockquote></details>

Good catch, this had gotten lost! Added it back.


<!-- Sent from Reviewable.io -->
